sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/StateFetchingIterators.java

```diff
@@ -318,23 +347,20 @@ public void append(List<T> values) {
       }
 
       // Combine all the individual blocks into one block containing all the values since
-      // they were mutated and we must evict all or none of the blocks. When consuming the blocks,
+      // they were mutated, and we must evict all or none of the blocks. When consuming the blocks,
       // we must have a reference to all or none of the blocks (which forces a load).
       List<Block<T>> blocks = existing.getBlocks();
-      long totalWeight = addBoundByMax(Caches.weigh(values), sumWeight(blocks));
       int totalSize = values.size();
-      for (int i = 0; i < blocks.size(); ++i) {
-        totalSize += blocks.get(i).getValues().size();
+      for (Block<T> block : blocks) {
+        totalSize += block.getValues().size();
       }
-      List<T> allValues = new ArrayList<>(totalSize);
-      for (int i = 0; i < blocks.size(); ++i) {
-        allValues.addAll(blocks.get(i).getValues());
+      WeightedList<T> allValues = new WeightedList<>(new ArrayList<>(totalSize), 0L);
+      for (Block<T> block : blocks) {
+        allValues.addAll(block.getValues(), block.getWeight());
       }
-      allValues.addAll(values);
+      allValues.addAll(values.getBacking(), values.getWeight());
```

Comment:
Would it makes sense to have an addAll(WeightedList) overload?

Reply:
Done, thanks!